### PR TITLE
Modify claim operation to 50 claims per tx

### PR DESCRIPTION
**Pull request by txhsl**
This PR modified claim gas operation to limited 50 claims per transaction. This may create more than 1 ClaimTransaction during 1 cmd but make it possible to claim gas from 10k transactions. Extra output is provided to show the progress.

### CODE REVIEW

**Comment by erikzhang** [comment]
Can you change it to send only one transaction at a time by default?

You can add a command line option:

```
claim gas [all]
```

Send one transaction only without `all`, otherwise send multiple.

**Comment by txhsl** [comment]
@erikzhang Sure, now it's an optional choice.

File: neo-cli/Shell/MainService.cs [hunk 1]
```diff
@@ -506,12 +506,33 @@ private bool OnClaimCommand(string[] args)
...
-                    ClaimTransaction tx = coins.Claim();
-                    if (tx != null)
+                    if (args.Length > 2)
                     {
-                        Console.WriteLine($"Tranaction Suceeded: {tx.Hash}");
+                        switch (args[2].ToLower())
+                        {
+                            case "all":
+                                ClaimTransaction[] txs = coins.ClaimAll();
+                                if (txs.Length > 0)
```
Comment (erikzhang) [inline neo-cli/Shell/MainService.cs]: The `ClaimAll()` method may return `null` and cause an `NullReferencException` on this line.

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo-cli/blob/13773f78ae7f7ccc78ae0aca1c7ff2f86a499f40/neo-cli/Shell/Coins.cs#L70

Shouldn't this line be:

```c#
Claims = claims.Take(MAX_CLAIMS_AMOUNT).ToArray(),
```
?

**Comment by txhsl** [comment]
:yum: , that's comprehensive.

File: neo-cli/Shell/Coins.cs [hunk 2]
```diff
@@ -87,6 +87,55 @@ public ClaimTransaction Claim()
...
+                Console.WriteLine($"no gas to claim");
+                return null;
+            }
+
+            CoinReference[] claims = current_wallet.GetUnclaimedCoins().Select(p => p.Reference).ToArray();
+            if (claims.Length == 0) return null;
+
+            using (Snapshot snapshot = Blockchain.Singleton.GetSnapshot())
+            {
+                int claim_count = claims.Length / MAX_CLAIMS_AMOUNT + 1;
```
Comment (erikzhang) [inline neo-cli/Shell/Coins.cs]: If `claims.Length == MAX_CLAIMS_AMOUNT`, `claim_count` should be 1.

File: neo-cli/Shell/Coins.cs [hunk 3]
```diff
@@ -87,6 +87,55 @@ public ClaimTransaction Claim()
...
+                        {
+                            new TransactionOutput
+                            {
+                                AssetId = Blockchain.UtilityToken.Hash,
+                                Value = snapshot.CalculateBonus(claims.Skip(i * MAX_CLAIMS_AMOUNT).Take(MAX_CLAIMS_AMOUNT)),
+                                ScriptHash = current_wallet.GetChangeAddress()
+                            }
+                        }
+                    };
+                    txs.SetValue((ClaimTransaction)SignTransaction(tx), i);
```
Comment (erikzhang) [inline neo-cli/Shell/Coins.cs]: `SignTransaction()` may return `null`, so this line should handle it, right?

File: neo-cli/Shell/Coins.cs [hunk 4]
```diff
@@ -128,10 +129,13 @@ public ClaimTransaction[] ClaimAll()
                             }
                         }
                     };
-                    txs.SetValue((ClaimTransaction)SignTransaction(tx), i);
+                    if ((tx = (ClaimTransaction)SignTransaction(tx)) != null)
```
Comment (erikzhang) [inline neo-cli/Shell/Coins.cs]: Should it break the loop if `tx` is `null`?

File: neo-cli/Shell/MainService.cs [hunk 5]
```diff
@@ -512,7 +512,7 @@ private bool OnClaimCommand(string[] args)
                         {
                             case "all":
                                 ClaimTransaction[] txs = coins.ClaimAll();
-                                if (txs != null && txs.Length > 0)
+                                if (txs.Length > 0)
```
Comment (txhsl) [inline neo-cli/Shell/MainService.cs]: I think `List<>` won't be `null`, I removed this judgement.

Comment (erikzhang) [reply neo-cli/Shell/MainService.cs]: I mean `tx`, not `txs`.

File: neo-cli/Shell/Coins.cs (hunk 4 above)
Comment (txhsl) [reply neo-cli/Shell/Coins.cs]: Stop the procession and leave the left unclaimed?

**Comment by txhsl** [comment]
@erikzhang maybe some change like that? https://github.com/neo-project/neo-cli/pull/236/commits/419ef1159cf388b030fe98f5a1cf981a6b30a499

**Pull request merged**